src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Logging/LogHubLoggerProviderFactoryBase.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Logging
+{
+    [Shared]
+    internal abstract class LogHubLoggerProviderFactoryBase : LogHubLoggerProviderFactory
+    {
+        // Unique, monotomically increasing ID to identify loghub session to persist
+        // across server restarts.
+        private static int _logHubSessionId;
+
+        private readonly RazorLogHubTraceProvider _traceProvider;
+        private readonly SemaphoreSlim _initializationSemaphore;
+        private DefaultLogHubLogWriter _currentLogWriter;
+
+        public LogHubLoggerProviderFactoryBase(RazorLogHubTraceProvider traceProvider)
+        {
+            if (traceProvider is null)
+            {
+                throw new System.ArgumentNullException(nameof(traceProvider));
+            }
+
+            _traceProvider = traceProvider;
+
+            _initializationSemaphore = new SemaphoreSlim(1, 1);
```

Comment:
Can we label these parameters for rubes like me who don't know what they do?

Reply:
Sure thing, yeah I agree it's a bit jarring just to see `1, 1`. For reference:

`new SemaphoreSlim(initialCount: 1, maxCount: 1);` are the args.